inspirehep/modules/workflows/utils.py

```diff
@@ -177,3 +180,63 @@ def download_file_to_workflow(workflow, name, url):
             req.raw.decode_content = True
             workflow.files[name] = req.raw
             return workflow.files[name]
+
+
+def store_root_json(record_uuid, source, json):
+    """Store the root json for a given source.
+
+    Given the ``record_uuid``, the ``source`` information (e.g. `arXiv` or
+    `Elsevier`) and the actual record in ``data`` store this information so
+    that it can be retrieved later.
+    """
+    with db.session.begin_nested():
+        record_source = WorkflowsRecordSources.query.filter(
+            WorkflowsRecordSources.record_id == record_uuid,
+            WorkflowsRecordSources.source == source
+        ).one_or_none()
+        if record_source is None:
+            record_source = WorkflowsRecordSources(
+                source=source,
+                json=json,
+                record_id=record_uuid
+            )
+        else:
+            record_source.json = json
+        db.session.add(record_source)
+
+
+def retrieve_root_json(record_uuid, source):
+    """Retrieve the root json for a given source.
+
+    Given a previously matched ``record_uuid``, the ``source`` information
+    (e.g. `arXiv` or `Elsevier`) returns the original record if existing or
+    empty json object otherwise.
+    """
+    entry = WorkflowsRecordSources.query.filter(
+        WorkflowsRecordSources.record_id == record_uuid,
+        WorkflowsRecordSources.source == source
+    ).one_or_none()
+    return entry.json if entry else {}
+
+
+def store_head_json(record_uuid, json):
+    """Store the head json for a given record.
+
+    Given the actual record in ``json`` and is key ``record_uuid``
+    store this information so that it can be retrieved later.
+    """
+    record = InspireRecord.get_record(record_uuid)
+    record.update(json)
```

Comment:
What are you doing with this `.update()`? if you are substituting the JSON that you should also call `.clear()` before. See also: https://github.com/inspirehep/inspire-next/commit/de83242fc2e05ef8eb644aef420cc8b6a0861e59